### PR TITLE
perf(dialog): reduce change detection cycles

### PR DESCRIPTION
**Pull request by arturovt**
This PR reduces change detection cycles for the dialog component by replacing
`HostListener` with `Renderer.listen` outside of the zone. The change detection should
be run only in specific cases, e.g. if the `esc` button has been clicked or the document
has been clicked and the target contains `dialog` class, but it shouldn't be run on each
button press and click event.
Note that the `HostListener` wraps the actual listener under the hood into the internal Angular
function which runs `markDirty()` before running the actual listener (the decorated class method).

Signed-off-by: Artur Androsovych <arthurandrosovich@gmail.com>

## Checklist

- [x] \*Added unit tests

### CODE REVIEW

File: projects/swimlane/ngx-ui/src/lib/components/dialog/dialog.component.ts [hunk 1]
```diff
@@ -151,15 +161,30 @@ export class DialogComponent implements OnInit, OnDestroy {
...
-    }
+  private setupEventListeners(): void {
+    this.ngZone.runOutsideAngular(() => {
+      const keydownEscListener = this.renderer.listen(this.element.nativeElement, 'keydown', (event: KeyboardEvent) => {
+        // We could've actually used `render.listener(element, 'keydown.esc')` and that would definitely work.
+        // Unfortunately, Angular delegates the event handling to the `KeyEventsPlugin` when it meets the `.` (dot)
+        // in the event name. The `KeyEventsPlugin` calls `ngZone.runGuarded` which triggers change detection anyway
+        // (no matter if we added event listeners in the root zone).
+        // In that case the change detection will be run only when the user clicks `esc` button.
+        if (event.keyCode === 27 && this.closeOnEscape) {
```
Comment (nartc) [inline projects/swimlane/ngx-ui/src/lib/components/dialog/dialog.component.ts:172]: nitpick: let's use `Keycodes.ESCAPE` from AngularCDK. When they deprecate Keycodes and switch over to Keys, we will be able to make the change here as well.

**Review by nartc — COMMENTED** [review]
LGTM. Just one small nitpick.

File: projects/swimlane/ngx-ui/src/lib/components/dialog/dialog.component.ts (hunk 1 above)
Comment (arturovt) [reply projects/swimlane/ngx-ui/src/lib/components/dialog/dialog.component.ts]: Thank you for bringing new knowledge! Updated the code.

**Comment by arturovt** [comment]
I will also update the changelog as you asked in another PR. I'll wait until another PR will get merged since I will have to resolve conflicts.

**Pull request closed without merging**